src/lib.rs

```diff
@@ -1,744 +1,345 @@
+                .funcs
+                .IRRootSignatureGetResourceCount(self.me as *const _);
+            let empty_location = IRResourceLocation {
+                resourceType: IRResourceType::IRResourceTypeInvalid,
+                space: 0,
+                slot: 0,
+                topLevelOffset: 0,
+                sizeBytes: 0,
+                resourceName: std::ptr::null(),
+            };
+            let mut resource_locations = vec![empty_location; n_resources];
+            self.funcs.IRRootSignatureGetResourceLocations(
+                self.me as *const _,
+                resource_locations.as_mut_ptr(),
+            );
+            resource_locations
+        }
+    }
 }
 
-#[repr(C)]
-#[derive(Copy, Clone)]
-pub union IRVersionedRootSignatureDescriptor_u {
-    // desc_1_0: IRRootSignatureDescriptor,
-    pub desc_1_1: IRRootSignatureDescriptor1,
+pub struct IRCompilerFactory {
+    funcs: Arc<bindings::metal_irconverter>,
 }
 
-#[repr(u32)]
-#[derive(Copy, Clone)]
-pub enum IRRootSignatureVersion {
-    IRRootSignatureVersion_1_0 = 0x1,
-    IRRootSignatureVersion_1_1 = 0x2,
-}
+impl IRCompilerFactory {
+    pub fn new(lib_path: impl AsRef<OsStr>) -> Result<Self, libloading::Error> {
+        let funcs = Arc::new(unsafe { bindings::metal_irconverter::new(lib_path)? });
+        Ok(Self { funcs })
+    }
 
-#[repr(C)]
-#[derive(Copy, Clone)]
-pub struct IRVersionedRootSignatureDescriptor {
-    pub version: IRRootSignatureVersion,
-    pub u: IRVersionedRootSignatureDescriptor_u,
-}
+    pub fn from_library(lib: libloading::Library) -> Result<Self, libloading::Error> {
+        let funcs = Arc::new(unsafe { bindings::metal_irconverter::from_library(lib)? });
+        Ok(Self { funcs })
+    }
 
-#[derive(Debug)]
-struct IRRootSignatureFn<'lib> {
-    create_from_descriptor: libloading::Symbol<
-        'lib,
-        unsafe extern "C" fn(
-            *const IRVersionedRootSignatureDescriptor,
-            *mut *mut IRErrorOpaque,
-        ) -> *mut IRRootSignatureOpaque,
-    >,
-    destroy: libloading::Symbol<'lib, unsafe extern "C" fn(*const IRRootSignatureOpaque) -> ()>,
-    get_resource_count: libloading::Symbol<'lib, unsafe extern "C" fn() -> ()>,
-    get_resource_locations: libloading::Symbol<'lib, unsafe extern "C" fn() -> ()>,
+    pub fn create_compiler(&self) -> IRCompiler {
+        let compiler = unsafe { self.funcs.IRCompilerCreate() };
+        IRCompiler {
+            funcs: self.funcs.clone(),
+            me: compiler,
+        }
+    }
 }
 
-pub struct IRRootSignature<'lib> {
-    me: *mut IRRootSignatureOpaque,
-    funcs: IRRootSignatureFn<'lib>,
+/// This object is not thread-safe, refer to [the Metal shader converter documentation], the "Multithreading considerations" chapter.
+///
+/// [the Metal shader converter documentation]: https://developer.apple.com/metal/shader-converter/
+pub struct IRCompiler {
+    funcs: Arc<bindings::metal_irconverter>,
+    me: *mut bindings::IRCompiler,
```

Comment:
```suggestion
    me: *mut bindings::IRCompiler,
    funcs: Arc<bindings::metal_irconverter>,
```

Let's be consistent for c-p purposes.